app/controllers/hats_controller.rb

```diff
@@ -62,4 +64,46 @@ def reject_request
 
     redirect_to "/hats/requests"
   end
+
+  def doff
+    @title = "Doffing a Hat"
+  end
+
+  def doff_by_user
+    if doff_with_reason(@hat, params[:reason])
+      redirect_to @user
+    else
+      redirect_to doff_hat_path(@hat)
+    end
+  end
+
+  private
+
+  def doff_with_reason(hat, reason)
+    if reason.present?
+      hat.doff_by_user_with_reason(@user, reason)
+      true
+    else
+      flash[:error] = "You must give a reason for the doffing."
+      false
+    end
+  end
+
+  def only_hat_user_or_moderator
+    if @hat.user == @user || @user.is_moderator?
+      true
+    else
+      redirect_to @user
+    end
+  end
+
+  def find_hat
```

Comment:
Please rename this to `find_hat!` and use `@user.hats.find(params[:id])`. This enforces that the user owns the hat and raises `RecordNotFound` for missing rather than requiring that you manage flash and return values.

Reply:
Though as soon as I read my own comment I realize you'll need this to be:

 `hats = @user.is_moderator? ? Hat.find(params[:id]) : @user...`